design/decisions/0004-entity-dregister-api.md

```diff
@@ -0,0 +1,147 @@
+
+The `mosquitto_sub` command provides options like `--remove-retained` to selectively remove all retained messages in a topic hierarchy by using topic filters.
+For example, the following command can be used to remove all retained messages under the `te/device/child01` topic tree:
+
+```sh
+mosquitto_sub --remove-retained -t "te/device/child01/+/+/#"
+```
+
+The mappers must also be updated to remove entities from their entity store on receipt of these clear messages.
+
+**Pros**
+
+* Existing out-of-the-box solution.
+
+**Cons**
+
+* Works only for topic schemes that ensures that parent and child entities have a common topic prefix.
+* Even the default topic scheme ensures common topic prefix only for a device and its services and not for its nested children.
+  So, it only works for simple deployments that does not have nested child devices.
+
+### Tedge deregister command
+
+Provide a `tedge` command that takes the topic id of the entity to be deregistered as follows:
+
+```sh
+tedge deregister te/device/child01//
+```
+
+This command subscribes to all entity messages, recreates the entity store on its own so that it can identify
+the given entity and all of its linked entities so that they can all be deregistered.
+
+:::note
+Subscribing to the given entity topic tree as `te/device/child01/#` isn't enough to identify all the linked entities
+as they may not share a common topic prefix.
+:::
+
+Once all the target entities are identified, it issues secondary subscriptions for all the data topics of these entities,
+like their twin data, alarm and command topics so that all those retained message are also cleared.
+On receipt of these clear messages, the mappers must clear the corresponding entries from their own in-memory entity stores,
+and propagate that to the cloud.
+
+The following options are provided to enable more fine-grained deregistration of entities:
+* `--services-only`: Delete only the associated services of the given device and not the device itself.
+* `--children-only`: Delete only the nested child devices not the device itself.
+* `--child-type`: Delete all the child devices with the given type
+
+**Pros**
+
+* Offers better targeting of entities, than what can be done using topic filters.
+
+**Cons**
+
+* Deregistrations can't be done remotely, unless `tedge` is installed on those remote device as well.
+
+### MQTT API
+
+Deregister entities by publishing MQTT requests as follows:
+
+```sh
+tedge mqtt pub te/device/child01///req/deregister '{"children-only": true}'
+```
+
+The `tedge-agent` handles these requests exactly as done by the `tedge deregister` command proposed above,
+by clearing all the retained messages associated with the target entities.
+The mappers react the same way as prosed in the previous solution.
+
+The responsibility of issuing the clear messages is left with the agent instead of the mapper,
+which already has its own entity store, to prevent multiple mappers from doing the same simultaneously.
```

Comment:
Yeah, multiple mappers issuing clear messages is not really a blocker issue or anything, considering their idempotent nature.

There is also the ownership aspect of that responsibility. If it's with `tedge-agent`, then we provide it out of the box irrespective of the mapper used. But, if we leave it to the mapper, then we need to replicate it in every mapper (of course we can reuse the logic with our own mappers), but even third-party mappers will have to support the same from their end. That's why I was more in favour of our own component owning that responsibility, and the mapper just being a mapper of those clear messages.

Reply:
> There is also the ownership aspect of that responsibility. If it's with `tedge-agent`, then we provide it out of the box irrespective of the mapper used. But, if we leave it to the mapper, then we need to replicate it in every mapper (of course we can reuse the logic with our own mappers), but even third-party mappers will have to support the same from their end. That's why I was more in favour of our own component owning that responsibility, and the mapper just being a mapper of those clear messages.

This makes sense. I wonder if this could be then pushed a bit further by also moving the auto-registration logic into the agent. The latter will be then responsible of maintaining a consistent view of the device entities.